[SYCL] Don't add `include/sycl/` to the system includes path

aelovikov-intel opened the pull request:
Fixes https://github.com/intel/llvm/issues/6770

Open question: I'm not sure if https://github.com/intel/llvm/tree/sycl/sycl/include/CL/__spirv should go into `include/CL/__spirv` or somewhere else. On the other hand, https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl.hpp has to be installed into `include/CL/sycl.hpp` per the specification.

sarnex reviewed (approved):
esimd changes lgtm, will let others review the overall change

mdtoguchi commented:
Driver changes look fine - seeing the number of header inclusion changes we have had to make internally - how do we minimize impact on users?

bader commented:
> Fixes #6770

Thanks for fixing this long-standing issue!

> On the other hand, https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl.hpp has to be installed into `include/CL/sycl.hpp` per the specification.

The spec says that 
> For compatibility with SYCL 1.2.1, SYCL provides another standard header file: <CL/sycl.hpp>, which can be included in place of <sycl/sycl.hpp>. In that case, all SYCL classes, constants, types and functions defined by this specification should exist within the ::cl::sycl C++ namespace.

It doesn't require implementation to put it into `include` directory. Moreover, it doesn't mandate the driver to add additional search paths for SYCL headers. It's DPC++ implementation choice. 

> I'm not sure if https://github.com/intel/llvm/tree/sycl/sycl/include/CL/__spirv should go into `include/CL/__spirv` or somewhere else.

What is your concern about the `__spirv` directory location?

aelovikov-intel commented:
> It doesn't require implementation to put it into `include` directory. Moreover, it doesn't mandate the driver to add additional search paths for SYCL headers. It's DPC++ implementation choice.

Yeah, I wasn't precise in what I said. What I meant is that `#include <CL/sycl.hpp>` must work and now that we don't put `include/sycl` in search path we have to move that header.

aelovikov-intel commented:
> What is your concern about the `__spirv` directory location?

It feels awkward to me that our implementation detail is located in OpenCL-specific folder. It might better be suited for `sycl/__spirv`.

aelovikov-intel commented:
> Driver changes look fine - seeing the number of header inclusion changes we have had to make internally - how do we minimize impact on users?

My personal opinion: we don't - it was a bug and now it's fixed.

bader commented:
> > What is your concern about the `__spirv` directory location?
> 
> It feels awkward to me that our implementation detail is located in OpenCL-specific folder. It might better be suited for `sycl/__spirv`.

Agree. SYCL-1.2.1 implementation put all headers into `CL` directory. When we added support for SYCL-2020, we moved all headers from `CL` to `sycl` directory. It sounds like it's a leftover of this work. I expect `CL` to have only one file - `sycl.hpp`, which includes `sycl/sycl.hpp` and imports all SYCL declarations to `cl` namespace.

aelovikov-intel commented:
> > > What is your concern about the `__spirv` directory location?
> > 
> > 
> > It feels awkward to me that our implementation detail is located in OpenCL-specific folder. It might better be suited for `sycl/__spirv`.
> 
> Agree. SYCL-1.2.1 implementation put all headers into `CL` directory. When we added support for SYCL-2020, we moved all headers from `CL` to `sycl` directory. It sounds like it's a leftover of this work. I expect `CL` to have only one file - `sycl.hpp`, which includes `sycl/sycl.hpp` and imports all SYCL declarations to `cl` namespace.

Good! I'll address it separately after this PR is merged in.

uditagarwal97 reviewed (approved):
SYCL RT changes LGTM

AlexeySachkov commented on sycl/include/sycl/accessor.hpp:
```diff
@@ -37,7 +37,7 @@
 #include <sycl/range.hpp>                             // for range
 #include <sycl/sampler.hpp>                           // for addressing_mode
-#include <ur_api.h>                                   // for UR_RESULT_ERRO...
+#include <sycl/ur_api.h>                              // for UR_RESULT_ERRO...
```
This change makes me think that `ur_api.h` is our own header, but that's not true. I wonder if we need to put it into some `_deps` subfolder for clarity.

aelovikov-intel replied:
I've been thinking about this too. I think even `sycl/_deps/ur_api.h` might be questionable, because one might want to use *both* UR and SYCL on the application side and I'd expect user's UR must somehow match the one `libsycl.so` was built with. Maybe I should be deploying these `ur_*.h/.hpp` at root `include/` directory instead.

Does anybody have any ideas here?

aelovikov-intel replied:
Implemented the move in the latest revision.

aelovikov-intel replied:
Or should it be `ur/ur_*.h` ? @intel/unified-runtime-reviewers

aelovikov-intel replied:
Ping @intel/unified-runtime-reviewers to answer the question above. @AlexeySachkov just in case as well.

aarongreig replied:
I don't think we ever intend for UR to be used directly by an application so we can eliminate that as a concern. In the UR repo the API header just lives in the root `include` so `ur/ur_api.h` would look a little weird from that perspective

aelovikov-intel replied:
I'm not sure if I prefer `_deps/ur_api.h` vs just `ur_api.h`, especially since the latter is the prior spelling of the includes. As such, `_deps/*` move should be in another PR, if at all.

joeatodd reviewed (approved):
SYCLcompat changes LGTM

aelovikov-intel commented:
Ping @intel/bindless-images-reviewers .

Seanst98 reviewed (approved):
Bindless LGTM
